src/contracts/middleware/BLSOperatorStateRetriever.sol

```diff
@@ -0,0 +1,134 @@
+
+        bytes memory quorumNumbers = BitmapUtils.bitmapToBytesArray(quorumBitmap);
+
+        return (quorumBitmap, getOperatorState(registryCoordinator, quorumNumbers, blockNumber));
+    }
+
+    /**
+     * @notice returns the ordered list of operators (id and stake) for each quorum. The AVS coordinator 
+     * may call this function directly to get the operator state for a given block number.s
+     * @param registryCoordinator is the registry coordinator to fetch the AVS registry information from
+     * @param quorumNumbers are the ids of the quorums to get the operator state for
+     * @param blockNumber is the block number to get the operator state for
+     * @return 2d array of operators. For each quorum, an ordered list of operators
+     */
+    function getOperatorState(IBLSRegistryCoordinatorWithIndices registryCoordinator, bytes memory quorumNumbers, uint32 blockNumber) public view returns(Operator[][] memory) {
+        IStakeRegistry stakeRegistry = registryCoordinator.stakeRegistry();
+        IIndexRegistry indexRegistry = registryCoordinator.indexRegistry();
+
+        Operator[][] memory operators = new Operator[][](quorumNumbers.length);
+        for (uint256 i = 0; i < quorumNumbers.length; i++) {
+            uint8 quorumNumber = uint8(quorumNumbers[i]);
+            bytes32[] memory operatorIds = indexRegistry.getOperatorListForQuorumAtBlockNumber(quorumNumber, blockNumber);
+            operators[i] = new Operator[](operatorIds.length);
+            for (uint256 j = 0; j < operatorIds.length; j++) {
+                bytes32 operatorId = bytes32(operatorIds[j]);
+                operators[i][j] = Operator({
+                    operatorId: operatorId,
+                    stake: stakeRegistry.getStakeForOperatorIdForQuorumAtBlockNumber(operatorId, quorumNumber, blockNumber)
+                });
+            }
+        }
+            
+        return operators;
+    }
+
+    /**
+     * @notice this is called by the AVS operator to get the relevant indices for the checkSignatures function
+     * if they are not running an indexer    
+     * @param registryCoordinator is the registry coordinator to fetch the AVS registry information from
+     * @param referenceBlockNumber is the block number to get the indices for
+     * @param quorumNumbers are the ids of the quorums to get the operator state for
+     * @param nonSignerOperatorIds are the ids of the nonsigning operators
+     * @return 1) the indices of the quorumBitmaps for each of given operators at the given blocknumber
+     *         2) the indices of the total stakes entries for the given quorums at the given blocknumber
+     *         3) the indices of the stakes of each of the nonsigners in each of the quorums they were a 
+     *            part of (for each nonsigner, an array of length the number of quorums they were a part of
+     *            that are also part of the provided quorumNumbers) at the given blocknumber
```

Comment:
question: if non-signers are/were only in a subset of quorums, how do we know which quorum each of the returned indices corresponds to?

Reply:
or does that not matter and get looked up automatically?

Reply:
gets looked up automatically. they are in ascending order of "1" bits in the operators quorum bitmap at the block number.